buchschloss/core.py

```diff
@@ -1097,6 +1099,218 @@ def view_str(member, *, login_context):
+        'name': the script name
+        'setlevel': the script's setlevel status ('-----' if not set)
+        'permissions': the scripts permissions, separated by ';'
+        """
+        return {
+            '__str__': str(script),
+            'name': script.name,
+            'setlevel': ('-----' if script.setlevel is None
+                         else utils.level_names[script.setlevel]),
+            'permissions': ';'.join(utils.get_name('Script::permissions::' + p.name)
+                                    for p in ScriptPermissions
+                                    if p in script.permissions)
+        }
+
+    @classmethod
+    @from_db(script=models.Script)
+    def execute(cls,
+                script: T.Union[str, models.Script],
+                function: T.Optional[str] = None,
+                *,
+                callbacks=None,
+                login_context):
+        """Execute a script
+
+        :param script: is a script name
+        :param function: is optionally a name of a function in the
+            script that takes no arguments. The function will be called.
+        :param callbacks: may be an alternative UI callback
+            dictionary to the default one
+        """
+        # avoid problems with circular import
+        # core -> lua.__init__ -> lua.objects -> core
+        from . import lua
+        if script.setlevel is not None:
+            script_lc_level = script.setlevel
+        else:
+            script_lc_level = login_context.level
+        script_lc = LoginType.SCRIPT(
+            script_lc_level, name=script.name, invoker=login_context)
+        ui_callbacks = callbacks or cls.callbacks
+        get_name_prefix = 'script-data::{}::'.format(script.name)
+        script_config = config.scripts.lua.get(script.name).mapping
+        if ScriptPermissions.STORE in script.permissions:
+            edit_func = partial(Script.edit, script.name, login_context=internal_priv_lc)
+            add_storage = (
+                lambda: Script.view_ns(script.name, login_context=internal_priv_lc).storage,
+                lambda data: edit_func(storage=data),
+            )
+        else:
+            add_storage = None
+        ns = lua.execute_script(
+            script.code,
+            script_lc,
+            add_ui=(None if ui_callbacks is None else (ui_callbacks, get_name_prefix)),
+            add_storage=add_storage,
+            add_requests=(ScriptPermissions.REQUESTS in script.permissions),
+            add_config=script_config,
+        )
+        if function is not None:
+            try:
+                ns[function]()
+            except Exception as e:
+                if config.debug:
+                    raise
+                logging.error('error executing script function: ' + str(e))
+                display = ':'.join((script.name, function))
+                raise BuchSchlossError('Script::execute', 'script_{}_exec_problem', display)
```

Comment:
This only displays errors when a function is configured. In some cases, problems are logged twice. How about moving all the error handling logic into here?